quickwit-search/src/export/leaf.rs

```diff
@@ -0,0 +1,127 @@
+// Quickwit
+//  Copyright (C) 2021 Quickwit Inc.
+//
+//  Quickwit is offered under the AGPL v3.0 and as commercial software.
+//  For commercial licensing, contact us at hello@quickwit.io.
+//
+//  AGPL:
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Affero General Public License as
+//  published by the Free Software Foundation, either version 3 of the
+//  License, or (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Affero General Public License for more details.
+//
+//  You should have received a copy of the GNU Affero General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::FastFieldCollectorBuilder;
+use crate::export::ExportSerializer;
+use crate::leaf::open_index;
+use crate::leaf::warmup;
+use crate::OutputFormat;
+use crate::SearchError;
+use quickwit_proto::LeafExportResult;
+use quickwit_storage::Storage;
+use std::sync::Arc;
+use tantivy::schema::Type;
+use tantivy::{query::Query, ReloadPolicy};
+use tokio_stream::wrappers::ReceiverStream;
+
+/// `leaf` step of export.
+// Note: we return a stream of a result with a tonic::Status error
+// to be compatible with the stream coming from the grpc client.
+// It would be better to have a SearchError but we need then
+// to process stream in grpc_adapater.rs to change SearchError
+// to tonic::Status as tonic::Status is required by the stream result
+// signature defined by proto generated code.
+pub async fn leaf_export(
+    query: Box<dyn Query>,
```

Comment:
I don't think we can pass a `Box<dyn Query>` here.
It assumes that all splits have the same schema.

It is a bummer but we want to parse the query for each split.
In the future it would be better to parse the query to an AST independent of the schema and pass that to each split.

Reply:
let's open a github issue to change that for `search` & `export` as we are doing that for the search part too.

Reply:
https://github.com/quickwit-inc/quickwit/issues/365